### DISCUSSION TITLE
Traffic shaping

### DISCUSSION THREAD

**Discussion by araujorm** [FoobarOy/foomuuri#58] [Ideas]
Traffic shaping (TC) in Linux is not exactly friendly to configure. Shorewall has support for it, and while it's still not too obvious how to make it work with some setups (e.g. when there is masquerade on the WAN interface), it helps quite a bit and it's nice to have it all integrated in a complex firewall setup - see https://shorewall.org/traffic_shaping.htm

Are there any plans to implement anything like this on foomuuri?

I think having something equivalent would greatly help people who are looking to migrate from Shorewall to consider foomuuri as the natural successor.

**Comment by kimheino** [comment]
Our current plan is to keep Foomuuri as firewall generator only. In other words, Foomuuri will generate nftables ruleset, and only that. (Foomuuri monitor is an exception to this.)

Traffic shaping / QoS requires two things:

1) Nftables rules to classify traffic with mark or priority. Foomuuri supports `mark_set` and I'll add `priority_set` to the next release.
2) `tc` rules to do the actual traffic shaping according to mark/priority.

Foomuuri can / will do part 1. For part 2 you need to find some other tool, or write a short shell script to setup `tc` rules.

Some links:

https://wiki.archlinux.org/title/Advanced_traffic_control
https://wiki.nftables.org/wiki-nftables/index.php/Classification_to_tc_structure_example

Are you sure you really need traffic shaping? I stopped doing that after I got 10 Mbps network connection. Basic rate limiting (mostly `saddr_rate` and `daddr_rate`) is enough for me.

**Comment by kimheino** [comment]
`priority_set` and `priority_match` are now added. I'll update wiki tomorrow.

**Comment by kimheino** [comment]
See also https://github.com/FoobarOy/foomuuri/discussions/67 for simple rate limit solution.
